network/data_link/at_parser.c

```diff
@@ -0,0 +1,1145 @@
+#include "at_parser.h"
+#include "error.h"
+#include "circular_buffer.h"
+#include "delay.h"
+#include "uart.h"
+#include "irq.h"
+#include "util.h"
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+
+/* Should be sizeof(async_msgs)/sizeof(*async_msgs) */
+#define NB_ASYNC_MESSAGES	1
+/* Should be sizeof(responses)/sizeof(*responses) */
+#define NB_RESPONSE_MESSAGES	3
+/* Max command length: at+cwsap=max_ssid_32,max_pass_64,0,0 -> 110 characters */
+#define CMD_BUFF_LEN		120u
+/* Maybe this could be smaller. Here must one response at a time */
+#define RESULT_BUFF_LEN		500u
+/* Used to remove warnings on strings */
+#define PUI8(X)			((uint8_t *)(X))
+/* Timeout waiting for module response. (20 seconds) */
+#define MODULE_TIMEOUT		20000
+/* Delay needed when executing reset command */
+#define RESET_DELAY		6000
+
+/******************************************************************************/
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+/* Command descriptor */
+struct cmd_desc {
+	struct at_buff	cmd;
+	uint32_t	type;
+};
+
+/* Structure that links the command string with the command macro by index */
+static const struct cmd_desc g_map[] = {
+	{{PUI8(""), 0}, AT_EXECUTE_OP},
+	{{PUI8("+RST"), 4}, AT_EXECUTE_OP},
+	{{PUI8("+GSLP"), 5}, AT_SET_OP},
+	{{PUI8("+GMR"), 4}, AT_EXECUTE_OP},
+	{{PUI8("+CWMODE"), 7}, AT_QUERY_OP | AT_SET_OP | AT_TEST_OP},
+	{{PUI8("+CWJAP"), 6}, AT_QUERY_OP | AT_SET_OP},
+	{{PUI8("+CWLAP"), 6}, AT_EXECUTE_OP},
+	{{PUI8("+CWQAP"), 6}, AT_EXECUTE_OP},
+	{{PUI8("+CWSAP"), 6}, AT_QUERY_OP | AT_SET_OP},
+	{{PUI8("+CWLIF"), 6}, AT_EXECUTE_OP},
+	{{PUI8("+CIPSTATUS"), 10}, AT_EXECUTE_OP},
+	{{PUI8("+CIPSTART"), 9}, AT_TEST_OP | AT_SET_OP},
+	{{PUI8("+CIPSEND"), 8}, AT_SET_OP},
+	{{PUI8("+CIPCLOSE"), 9}, AT_EXECUTE_OP | AT_SET_OP},
+	{{PUI8("+CIFSR"), 6}, AT_EXECUTE_OP},
+	{{PUI8("+CIPMUX"), 7}, AT_QUERY_OP | AT_SET_OP},
+	{{PUI8("+CIPSERVER"), 10}, AT_SET_OP},
+	{{PUI8("+CIPMODE"), 8}, AT_QUERY_OP | AT_SET_OP},
+	{{PUI8("+CIPSTO"), 7}, AT_QUERY_OP | AT_SET_OP},
+	{{PUI8("+PING"), 5}, AT_SET_OP}
+};
+
+/* Structure storing a circular buffer for a connection */
+struct at_circular_buff {
+	/* Size of the buffer */
+	uint32_t	size;
+	/* Address to the buffer */
+	uint8_t		*buff;
+	/* Read index */
+	uint32_t	ri;
+	/* Write index */
+	uint32_t	wi;
+	/* Number of current elements in buffer */
+	uint32_t	count;
+	/* Last read len */
+	uint32_t	last_len;
+	/* Set if overflow */
+	bool		is_overflow;
+	/* Used for polling in blocking read */
+	bool		new_data;
+};
+
+/* Structure storing a connection status */
+struct connection_desc {
+	/* Connection buffer */
+	volatile struct at_circular_buff	cbuff;
```

Comment:
why is this volatile ? it does not contain register information, loading data into the buffers and index is still done by the application. And also we are not in a multi threaded scenario.

Reply:
Because the values are modified inside and outside the uart interrupt.